src/core/ui/egl/rpi-input.cpp

```diff
@@ -0,0 +1,221 @@
+#if OTTO_UI_EGL
+
+#include <fcntl.h>
+#include <linux/input.h>
+#include <plog/Log.h>
+#include <string>
+#include <vector>
+#include "core/ui/mainui.hpp"
+#include "filesystem.hpp"
+
+namespace otto::ui {
+  static auto constexpr RELEASE = 0;
+  static auto constexpr PRESS   = 1;
+  static auto constexpr REPEAT  = 2;
+
+  namespace fs = otto::filesystem;
+
+  static bool ends_with(const std::string& s, const std::string& suffix)
+  {
+    return s.rfind(suffix) == (s.size() - suffix.size());
+  }
+
+  static int open_device(const std::string& device_type)
+  {
+    auto path = fs::path("/dev/input/by-id");
+
+    for (const auto& entry : fs::directory_iterator(path)) {
+      auto file = entry.path().string();
+      if (ends_with(file, device_type)) {
+        auto fullpath = path / file;
+        auto fd       = open(fullpath.c_str(), O_RDONLY | O_NONBLOCK);
+        auto result   = ioctl(fd, EVIOCGRAB, 1);
+        if (result != 0) {
+          LOGE << "Couldn't get exclusive input access to "
+               << fullpath.string();
+        }
+
+        return fd;
+      }
+    }
+
+    return -1;
+  }
+
+  std::unique_ptr<std::vector<input_event>> read_events(int device)
```

Comment:
what exactly is the reason to not return the vector directly?

Reply:
I went back and forth on this.  I'm fine with simply returning a vector here -- it does mean a copy construction, however.  Perhaps the compiler would optimize that away?

Reply:
Return value optimization has been a thing for a long time i think, and since C++11 that would be moved otherwise anyway.

Since c++17, returning from a function is guarrantied to elide the copy. There is a great talk by Herb sutter from cppcon to catch up on the modern guidelines for value vs reference semantics

Reply:
Good to know.  I'll make the change.